src/tools/tests/schema2kotlin-test.ts

```diff
@@ -236,10 +267,181 @@ describe('schema2kotlin', () => {
+    ) : EntityBase("Thing", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates variable entity with private constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads ~a with {num: Number}
+         `,
+      `@Suppress("UNCHECKED_CAST")
+    class T_H1 private constructor(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : EntityBase("T_H1", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates copy and mutate by entity fields', async () => await assertCopyMethods(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = Thing(num = num)
+        /**
+         * Use this method to create a new version of an existing entity.
+         * Storing the mutation will overwrite the existing entity in the set, if it exists.
+         */
+        fun mutate(num: Double = this.num) = Thing(
+            num = num,
+            entityId = entityId,
+            creationTimestamp = creationTimestamp,
+            expirationTimestamp = expirationTimestamp
+        )`
+    ));
+    // TODO(alxr): Why do we omit the docstring for Wasm?
+    it('generates only copy method by entity fields for wasm', async () => await assertCopyMethodsForWasm(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `
+        fun copy(num: Double = this.num) = Thing(num = num)
+        `
+    ));
+    it('copies underlying data for type variables.', async () => await assertCopyMethods(
+      `particle T
+         h1: reads ~a with Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = Thing(num = num)
+            .also { this.copyInto(it) }
+        /**
+         * Use this method to create a new version of an existing entity.
+         * Storing the mutation will overwrite the existing entity in the set, if it exists.
+         */
+        fun mutate(num: Double = this.num) = Thing(
+            num = num,
+            entityId = entityId,
+            creationTimestamp = creationTimestamp,
+            expirationTimestamp = expirationTimestamp
+        ).also { this.copyInto(it) }`
+    ));
+    it('generates only copy method for variables for wasm', async () => await assertCopyMethodsForWasm(
+      `particle T
+         h1: reads  ~a with Thing {num: Number}`,
+      `
+        fun copy(num: Double = this.num) = Thing(num = num)
```

Comment:
I want to highlight a change: For variables in the Wasm case, we DO NOT copy the base entity. 
WasmEntity !== EntityBase. 

WDYT? @piotrswigon

Reply:
Right now both (wasm and non wasm) are the same, but also we don't do anything for copy(), so our feature will not work. We need to add something to fix this.